### DISCUSSION TITLE
Delete command

### DISCUSSION THREAD

**Discussion by TomWright** [TomWright/dasel#64] [Ideas]
Dasel doesn't currently support deletions.

## Idea

A new `dasel delete` command that takes a selector and deletes it.

It should handle the same kinds of flags as `dasel put` except for the fact you don't need to provide any variable type or value.

E.g.
```
$ echo '{"name": "Tom", "age": 27}' | dasel delete -p json .name 
{
  "age": 27
}
```

```
$ echo '[1,2,3,4]' | dasel delete -p json .[1] 
[
  1,
  3,
  4
]
```

## Unknowns

1. Should `ValueNotFound` errors be thrown if the value you're trying to delete doesn't exist?
1. Should errors be thrown if the path to the value you're trying to delete doesn't exist?
1. What happens if you delete the root node?
   1. Return a blank file?
   1. Return `null`?
   1. Throw an error?

**Comment by link4531** [comment]
1. ValueNotFound should occur just like 'rm non-existent.file' would :)
2. I agree an error should be throw just like select
3. well as long as there is a way to replace an existing node with dasel put string , where we dont have to remove and insert. this should be fine. otherwise you would need a flag for non recursive deletion of a parent node without removal of its children. < which could obviously break the file structure but I dont know...

**Reply by TomWright** [reply]
Agreed with this - although there is no way non non-recursive deletion since when the parent is deleted all of the children must also be deleted.
You could use search selectors and such to target specific nodes, but yes I agree in general.

This is definitely on my to-do list, I just haven't gotten around to it yet.

**Comment by TomWright** [comment]
I'm working on this here: https://github.com/TomWright/dasel/pull/145

**Comment by TomWright** [comment]
Just a couple more tests to add to increase coverage but it's working and the docs can be found here: https://daseldocs.tomwright.me/usage/delete

There will be a release soon.

**Comment by TomWright** [comment]
The delete command is now released under `v1.16.0` and should be on homebrew shortly.

Docs are here: https://daseldocs.tomwright.me/usage/delete

Quick example:
```bash
echo '{
  "email": "contact@tomwright.me",
  "name": "Tom"
}' | dasel delete -p json '.email' 'contact@tomwright.me'
{
  "name": "Tom"
}
```

**Comment by link4531** [comment]
Oh this is a good day.
